sql/idempotent/011-vectorizer-api.sql

```diff
@@ -0,0 +1,553 @@
+        , queue_schema
+        , queue_table
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- create trigger on source table to populate queue
+    perform ai._vectorizer_create_source_trigger
+    ( _trigger_name
+    , queue_schema
+    , queue_table
+    , _source_schema
+    , _source_table
+    , _source_pk
+    );
+
+    -- create view
+    perform ai._vectorizer_create_view
+    ( view_schema
+    , view_name
+    , _source_schema
+    , _source_table
+    , _source_pk
+    , target_schema
+    , target_table
+    );
+
+    -- grant select view to _grant_to roles
+    if grant_to is not null then
+        select pg_catalog.format
+        ( $sql$grant select on %I.%I to %s$sql$
+        , view_schema
+        , view_name
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- schedule the async ext job
+    select ai._vectorizer_schedule_async_ext_job
+    (_vectorizer_id
+    , scheduling
+    ) into _job_id
+    ;
+    if _job_id is not null then
+        scheduling = pg_catalog.jsonb_insert(scheduling, array['job_id'], to_jsonb(_job_id));
+    end if;
+
+    insert into ai.vectorizer
+    ( id
+    , source_schema
+    , source_table
+    , source_pk
+    , target_schema
+    , target_table
+    , view_schema
+    , view_name
+    , trigger_name
+    , queue_schema
+    , queue_table
+    , config
+    )
+    values
+    ( _vectorizer_id
+    , _source_schema
+    , _source_table
+    , _source_pk
+    , target_schema
+    , target_table
+    , view_schema
+    , view_name
+    , _trigger_name
+    , queue_schema
+    , queue_table
+    , pg_catalog.jsonb_build_object
+      ( 'version', '@extversion@'
+      , 'embedding', embedding
+      , 'chunking', chunking
+      , 'indexing', indexing
+      , 'formatting', formatting
+      , 'scheduling', scheduling
+      )
+    );
+
+    -- grant select on vectorizer table _grant_to roles
+    -- TODO: is there a better way to do this?
+    if grant_to is not null then
+        -- grant select on the users that do NOT already have it
+        for _sql in
+        (
+            select pg_catalog.format($sql$grant select on ai.vectorizer to %I$sql$, x)
+            from pg_catalog.unnest(grant_to) x
+            where not pg_catalog.has_table_privilege
+            ( x
+            , 'ai.vectorizer'
+            , 'select'
+            )
+        )
+        loop
+            execute _sql;
+        end loop;
+    end if;
+
+    -- insert into queue any existing rows from source table
+    select pg_catalog.format
+    ( $sql$
+    insert into %I.%I (%s)
+    select %s
+    from %I.%I x
+    ;
```

Comment:
Should we put this behind a flag like `embed_existing_data bool default true`?

Reply:
95099edd1a6d7dd141edb35cb8c295629787b3ba